### PR TITLE
[Whisper] Computing features on GPU in batch mode for whisper feature extractor.   

### LINKED ISSUE #29901: [Whisper] Computing features on CPU slows down WhisperFeatureExtractor
### Feature request

In the WhisperFeatureExtractor class, to reduce the time it takes to compute audio features, an option should be added to compute the features on GPU in batch mode.

### Motivation

Currently, the Whisper Feature Extractor processes audio features on the CPU one by one for each batch, which slows down the process. Additionally, it doesn't utilize GPU operations despite using torch functions.

We ( + @vaibhavagg303 ) modified the code to enable batch feature extraction on CPU, cutting the average time for batches of 8 from 1.5 seconds to 0.25 seconds — depending on available CPU resources. Furthermore, we transferred all variables — mel filters and hanning window — in the function __torch_extract_fbank_features_  to GPU, further reducing **the average time to 0.02 seconds — a significant reduction from 1.5 seconds.**

Whisper Models, already with high latency, experience even greater latency due to processing input features on the CPU without batching. By doing batch-wise inference on GPU, the total inference time for Whisper-Small is reduced significantly, **bringing it down by whopping 25%.** 

@sanchit-gandhi 

The code used to measure the time taken is as follows. 
```
from transformers import WhisperProcessor, WhisperForConditionalGeneration
feature_extractor = WhisperFeatureExtractor.from_pretrained("openai/whisper-small")
import librosa
import time

processor = WhisperProcessor.from_pretrained("openai/whisper-small", language='english')
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-small")
model.config.forced_decoder_ids = None
model = model.to('cuda')

total_time = 0
batch_size= 8

for i in range(0, len(df), batch_size):
  sample = df[i:i+ batch_size]
  files = sample['file'].tolist()
  sample_array = []
  for file in files:
    data, sample_rate = librosa.load(file, sr = 16000)
    sample_array.append(data)
 
 st = time.time()
  input_features = feature_extractor(sample_array, sampling_rate=16000).input_features
  input_features = torch.tensor(input_features).to('cuda')

  with torch.no_grad():
    predicted_ids = model.generate(input_features, language='english')
  transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)

  total_time += time.time() - st
```

### Your contribution

This is the PR by @vaibhavagg303.
https://github.com/huggingface/transformers/pull/29900

### PR DESCRIPTION
**Pull request by vaibhavagg303**
# What does this PR do?

This pull request introduces a feature enabling the computation of audio features for the Whisper model in batch mode on the GPU. This enhancement substantially reduces latency during feature extraction, thereby enhancing overall performance.
cc: @yashjogi

@sanchit-gandhi @ArthurZucker @hollance  Please look into it.

fixes #29901

### CODE REVIEW

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 1]
```diff
@@ -270,9 +292,12 @@ def __call__(
...
 
         extract_fbank_features = (
-            self._torch_extract_fbank_features if is_torch_available() else self._np_extract_fbank_features
+            self._torch_extract_fbank_features_batch if is_torch_available() else self._np_extract_fbank_features
         )
-        input_features = [extract_fbank_features(waveform) for waveform in input_features[0]]
+        if is_torch_available():
+            input_features = extract_fbank_features(input_features[0], device)
+        else:
+            input_features = [extract_fbank_features(waveform) for waveform in input_features[0]]
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/feature_extraction_whisper.py:300]: I don't think we need the extra `is_torch_available`

Comment (ArthurZucker) [reply src/transformers/models/whisper/feature_extraction_whisper.py:300]: whether or not we batch should be handle in `_torch_extract_fbank_features`.

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 2]
```diff
@@ -131,6 +131,27 @@ def _torch_extract_fbank_features(self, waveform: np.array) -> np.ndarray:
         log_spec = (log_spec + 4.0) / 4.0
         return log_spec.numpy()
 
+    def _torch_extract_fbank_features_batch(self, waveforms: np.array, device: str) -> np.ndarray:
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/feature_extraction_whisper.py:134]: I don't think we need a new function we can update the original one no?

**Review by ArthurZucker — COMMENTED** [review]
Very nice addition!

File: src/transformers/models/whisper/feature_extraction_whisper.py (hunk 2 above)
Comment (vaibhavagg303) [reply src/transformers/models/whisper/feature_extraction_whisper.py]: @ArthurZucker,
I've committed the recent changes for your review. Please take a look.

Additionally, I've observed a slight discrepancy in the results when computing features on the GPU. The mismatch is on the order of 1e-5 and is primarily attributed to differences in the Short-Time Fourier Transform (STFT) computation results. Can you suggest what should I do here?

Comment (ArthurZucker) [reply src/transformers/models/whisper/feature_extraction_whisper.py]: it's alright IMO, as long as it translates to 1e-5 in the output logits!

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 3]
```diff
@@ -94,41 +94,57 @@ def __init__(
             mel_scale="slaney",
         )
 
-    def _np_extract_fbank_features(self, waveform: np.array) -> np.ndarray:
+    def _np_extract_fbank_features(self, waveform_batch: np.array) -> np.ndarray:
         """
         Compute the log-mel spectrogram of the provided audio, gives similar results to Whisper's original torch
         implementation with 1e-5 tolerance.
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/feature_extraction_whisper.py:100]: let's update the doc here!

**Review by ArthurZucker — COMMENTED** [review]
Looks good. Let's add a test the uses `require_torch_gpu` to make sure this runs on GPU!

File: src/transformers/models/whisper/feature_extraction_whisper.py (hunk 3 above)
Comment (vaibhavagg303) [reply src/transformers/models/whisper/feature_extraction_whisper.py:100]: I have added some changes, but I'm unsure if it's correct. @ArthurZucker

**Review by ArthurZucker — APPROVED** [review]
LGTM, let's ask for @sanchit-gandhi 's approval as well here!

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 4]
```diff
@@ -94,41 +94,58 @@ def __init__(
...
+        log_spec_batch = np.array(log_spec_batch)
+        return log_spec_batch
 
     def _torch_extract_fbank_features(self, waveform: np.array) -> np.ndarray:
         """
-        Compute the log-mel spectrogram of the provided audio using the PyTorch STFT implementation.
+        Compute the log-mel spectrogram of the audio using PyTorch's GPU-accelerated STFT implementation with batching,
+        yielding results similar to cpu computing with 1e-5 tolerance.
         """
+        device = "cuda" if torch.cuda.is_available() else None
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/feature_extraction_whisper.py:125]: In Transformers, we leave it to the user to set the GPU accelerator device. E.g. when we load the Whisper model, we manually move it to the GPU, rather than having Transformers do it automatically:
```python
import torch
from transformers import AutoModelForSpeechSeq2Seq

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "distil-whisper/distil-large-v3"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)
```
Whereas here, we set the GPU accelerator device in the feature extraction code, which is a divergence from the Transformers philosophy.

Could we allow the user to pass an argument `device` to the `__call__` method of the Whisper feature extractor that lets them set the appropriate torch device for the computation?

File: tests/models/whisper/test_feature_extraction_whisper.py [hunk 5]
```diff
@@ -207,6 +207,7 @@ def _load_datasamples(self, num_samples):
 
         return [x["array"] for x in speech_samples]
 
+    @require_torch_gpu
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_feature_extraction_whisper.py:210]: Thanks for updating this test! Could we also add one test that checks that we get the correct values from the torch feature extractor with **batched** inputs? Since this is a new feature we're getting from the changes proposed above.

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for this PR @vaibhavagg303 - performance improvements to Whisper are always most welcome! In general, I'm in favour of the changes towards faster feature extractors. Do you have any numbers on how much faster this CUDA variant is? My only reluctancy in merging this PR would be if the performance gains are negligible, but we add extra complexity due to the potential new `device` argument. To check this, you can adapt and update the toy benchmark that was proposed as part of the first torch stft addition: https://github.com/huggingface/transformers/pull/26119#issue-1892916324. Otherwise, I've left some small suggestions below.

**Comment by sanchit-gandhi** [comment]
cc @kamilakesbi for viz

File: src/transformers/models/whisper/feature_extraction_whisper.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/whisper/feature_extraction_whisper.py:125]: Agreed here, but it's unintuitive, since it's a device that won't store a model, I brushed it as being alright. We don't move weights to device but move the processing. However if sft is supported in other frameworks, having the arg is better

File: tests/models/whisper/test_feature_extraction_whisper.py (hunk 5 above)
Comment (yashjogi) [reply tests/models/whisper/test_feature_extraction_whisper.py:210]: https://github.com/huggingface/transformers/issues/29901
The details regarding the time improvement are mentioned in this issue. It reduces the complete inference time for the Whisper transcript by around 25%. And we also see a significant decrease in extracting features from 1.5 seconds ( without batch and CPU ) to 0.01 seconds ( by using batch and GPU ), which is a significant reduction.
@sanchit-gandhi

**Comment by yashjogi** [comment]
https://github.com/huggingface/transformers/issues/29901
The details regarding the time improvement are mentioned in this issue. It reduces the complete inference time for Whisper-small by around 25%. And we also see a significant time decrease in extracting features from 1.5 seconds ( without batch and CPU ) to 0.01 seconds ( by using batch and GPU ).
@sanchit-gandhi

**Comment by sanchit-gandhi** [comment]
Thanks for the benchmark link @vaibhavagg303, that's most helpful! Once follow-up question to your benchmark: you mention that torch gpu stft "cuts the average time for batches of 8 from 1.5 seconds to 0.25 seconds". How does the compute time change for bsz=1, since this is also a common use-case for computing features in short-form mode (<30s).

**Comment by sanchit-gandhi** [comment]
As a follow-up PR: we'll need to update the ASR pipeline class to compute batched input features, since currently it uses bsz=1 always for the feature extraction
https://github.com/huggingface/transformers/blob/9b0a8ea7d1d6226b76cfdc645ce65e21157e2b50/src/transformers/pipelines/automatic_speech_recognition.py#L66

**Comment by yashjogi** [comment]
> Thanks for the benchmark link @vaibhavagg303, that's most helpful! Once follow-up question to your benchmark: you mention that torch gpu stft "cuts the average time for batches of 8 from 1.5 seconds to 0.25 seconds". How does the compute time change for bsz=1, since this is also a common use-case for computing features in short-form mode (<30s).

This is the computation time for **8 audios.**
| Mode | Time |
|-------|------|
| CPU Batch Size 1 |  1.5 seconds |
| CPU Batch Size 8 |  0.25 seconds |
| GPU Batch Size 8 | 0.02 seconds |

**Comment by sanchit-gandhi** [comment]
Thanks @vaibhavagg303. To clarify my question above, what's the speed up of `CPU Batch Size 1` vs `GPU Batch Size 1`?

**Comment by vaibhavagg303** [comment]
> Thanks @yashjogi. To clarify my question above, what's the speed up of `CPU Batch Size 1` vs `GPU Batch Size 1`?

It's around `0.02291` for `GPU Batch Size 1` @sanchit-gandhi

**Comment by yashjogi** [comment]
Also, we found out that this is a huge bottleneck for training Whisper -- this simple change reduced the training time by almost 9 times, depending on the CPU resources of the machine we're training on. This means that it would decrease the time to train distil-whisper as well! https://github.com/huggingface/distil-whisper/blob/main/training/run_distillation.py
@sanchit-gandhi

File: src/transformers/models/whisper/feature_extraction_whisper.py (hunk 4 above)
Comment (vaibhavagg303) [reply src/transformers/models/whisper/feature_extraction_whisper.py:125]: @sanchit-gandhi I have added this previously in my commits, can you please verify if this meets your requirements?

File: tests/models/whisper/test_feature_extraction_whisper.py [hunk 6]
```diff
@@ -253,3 +255,37 @@ def test_zero_mean_unit_variance_normalization_trunc_np_longest(self):
 
         self.assertTrue(np.all(np.mean(audio) < 1e-3))
         self.assertTrue(np.all(np.abs(np.var(audio) - 1) < 1e-3))
+
+    @require_torch_gpu
+    @require_torch
+    def test_torch_integration_batch(self):
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_feature_extraction_whisper.py:261]: Nice test!

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 7]
```diff
@@ -94,41 +94,57 @@ def __init__(
...
-            mel_filters=self.mel_filters,
-            log_mel="log10",
-        )
-        log_spec = log_spec[:, :-1]
-        log_spec = np.maximum(log_spec, log_spec.max() - 8.0)
-        log_spec = (log_spec + 4.0) / 4.0
-        return log_spec
-
-    def _torch_extract_fbank_features(self, waveform: np.array) -> np.ndarray:
+        log_spec_batch = []
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/feature_extraction_whisper.py:102]: Should we throw an error here if `device!="cpu"` and we're in the numpy branch? 
```suggestion
        if device != "cpu":
                raise ValueError(f"Got device `{device}` for feature extraction, but feature extraction on CUDA accelerator devices requires torch, which is not installed. Either set `device='cpu'`, or install torch according to the official instructions: https://pytorch.org/get-started/locally/")
        log_spec_batch = []
```

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 8]
```diff
@@ -165,6 +181,7 @@ def __call__(
         max_length: Optional[int] = None,
         sampling_rate: Optional[int] = None,
         do_normalize: Optional[bool] = None,
+        device: Optional[str] = "cpu",
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/feature_extraction_whisper.py:184]: Could you add this argument to the args docstring, explaining a bit what it does?

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for the updates @vaibhavagg303 - just one small suggestion about the docstring, otherwise LGTM!

**Comment by vaibhavagg303** [comment]
Hi @ArthurZucker, can you please review and approve these changes?

**Review by ArthurZucker — APPROVED** [review]
Thanks for getting through this and bringing fast whisper processing! 🤗

**Comment by sanchit-gandhi** [comment]
Thanks for your contribution @vaibhavagg303!
